### PR TITLE
APNS - add support for HTTP/2 interface for APNS

### PR DESCRIPTION
**Pull request by soulfly**
Add support for HTTP/2 interface for APNS
Created a separate service for this - **apns2**

APNS HTTP/2 protocol overview https://developer.apple.com/library/content/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/APNSOverview.html#//apple_ref/doc/uid/TP40008194-CH8-SW1

### CODE REVIEW

**Comment by amaierhofer** [comment]
Nice work @soulfly @Nattfodd, this looks like a big step forward.

Could you please remove `ruby-2.0.0` from `.travis.yml` to make the build pass? Another thing I noticed is that other deliveries (`adm, gcm, wpns`...) use a dedicated `handle_response` method. It is probably a good idea to stick to this pattern. Also I like `wpns` wordy error messages. What do you think?

**Comment by Nattfodd** [comment]
I don't think it's a good idea to hardcode error messages, there're [plenty of them](https://developer.apple.com/library/content/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/CommunicatingwithAPNs.html#//apple_ref/doc/uid/TP40008194-CH11-SW17). In most cases shortened reasons that Apple sends in response (i.e. `BadDeviceToken`, `TooManyProviderTokenUpdates`) say for themselves.

For the rest things, I agree with you, changes are on the way

File: lib/rpush/daemon/apns2/delivery.rb [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+        def handle_response(response)
+          code = response.code
+          case code
+          when 200
+            ok(response)
+          when *RETRYABLE_CODES
+            service_unavailable(response)
+          else
+            reflect(:notification_id_failed,
```
Comment (amaierhofer) [inline lib/rpush/daemon/apns2/delivery.rb]: handled by `Rpush::Daemon::Delivery`

File: spec/functional/apns2_spec.rb [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+        {
+          ":status" => "500",
+          "apns-id"=>"C6D65840-5E3F-785A-4D91-B97D305C12F6"
+        }
+      }
+
+      it 'reflects :notification_id_will_retry' do
+        Rpush.reflect do |on|
+          on.notification_id_will_retry do |app, id, timer|
+            expect(app).to be_kind_of(Rpush::Client::ActiveRecord::Apns2::App)
```
Comment (amaierhofer) [inline spec/functional/apns2_spec.rb]: Fails with redis, please change this to general class `Rpush::Apns2::App`

File: spec/functional/apns2_spec.rb [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        expect do
+          Rpush.push
+          notification.reload
+        end.to change(notification, :retries).to eq(1)
+      end
+
+      it 'reflects :notification_id_will_retry' do
+        Rpush.reflect do |on|
+          on.notification_id_will_retry do |app, id, timer|
+            expect(app).to be_kind_of(Rpush::Client::ActiveRecord::Apns2::App)
```
Comment (amaierhofer) [inline spec/functional/apns2_spec.rb]: Please change this to general class `Rpush::Apns2:App`

File: lib/rpush/daemon/dispatcher/apns_http2.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+          @client = NetHttp2::Client.new(url,
+            ssl_context:     prepare_ssl_context,
+            connect_timeout: DEFAULT_TIMEOUT)
+        end
+
+        def dispatch(payload)
+          @delivery_class.new(@app, @client, payload.batch, payload.notification).perform
+        end
+
+        def cleanup
```
Comment (amaierhofer) [inline lib/rpush/daemon/dispatcher/apns_http2.rb]: Does http2 offer something similar to `@http.shutdown` called in `Push::Daemon::Dispatcher::Http`?

File: lib/rpush/daemon/apns2/delivery.rb [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+          mark_retryable(@notification, Time.now + 10.seconds)
+          # Logs should go last as soon as we need to initialize
+          # retry time to display it in log
+          failed_message_to_log(response)
+          retry_message_to_log
+        end
+
+        def do_post
+          request = prepare_request
+          resp = @client.call(:post, request[:path],
```
Comment (daniel-illi) [inline lib/rpush/daemon/apns2/delivery.rb]: wouldn't it make sense to use #call_async to take advantage of streaming the requests?

File: spec/functional/apns2_spec.rb [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+    context 'shit' do
+      let(:fake_http_resp_headers) {
+        {
+          ":status" => "500",
+          "apns-id"=>"C6D65840-5E3F-785A-4D91-B97D305C12F6"
+        }
+      }
+
+      it 'reflects :notification_id_will_retry' do
```
Comment (amaierhofer) [inline spec/functional/apns2_spec.rb]: Could this be included in the previous context?

File: spec/functional/apns2_spec.rb [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+          "apns-id"=>"C6D65840-5E3F-785A-4D91-B97D305C12F6"
+        }
+      }
+
+      it 'fails but retries delivery several times' do
+        notification = create_notification
+        expect do
+          Rpush.push
+          notification.reload
+        end.to change(notification, :retries).to eq(1)
```
Comment (amaierhofer) [inline spec/functional/apns2_spec.rb]: Please avoid checking exact number, e.g. use `change(notification, :retries)`. Reason is that with redis this number increments to 2, but this is not an issue with your code.

File: lib/rpush/daemon/dispatcher/apns_http2.rb (hunk 4 above)
Comment (Nattfodd) [inline lib/rpush/daemon/dispatcher/apns_http2.rb]: It does, thanks for your review!

File: lib/rpush/daemon/apns2/delivery.rb (hunk 5 above)
Comment (Nattfodd) [inline lib/rpush/daemon/apns2/delivery.rb]: Not sure if we win something with streaming requests in this particular case, the case is very simple and request/response is pretty short, but for the future it makes sense. Check if everything is fine please, and thanks for review!

File: lib/rpush/daemon/dispatcher/apns_http2.rb (hunk 4 above)
Comment (soulfly) [inline lib/rpush/daemon/dispatcher/apns_http2.rb]: @amaierhofer I gave you an access to our PR

**Comment by soulfly** [comment]
Does anybody have more comments here?

How we can merge it into core and release new version?

**Comment by soulfly** [comment]
@aried3r, @amaierhofer, @ileitch guys could you merge this?

**Comment by amaierhofer** [comment]
Thanks for all your work on this. I have verified functionality via simple integration tests and it worked :) Thanks again @soulfly, @Nattfodd  and @daniel-illi

**Comment by aried3r** [comment]
I Ruby 2.0.0 support being dropped? I'm all for it, but that is a backwards incompatible change which would need a v3.

**Comment by amaierhofer** [comment]
@aried3r  agreed. I have bumped the version 3.0.0

**Comment by Fivell** [comment]
@soulfly,  rpush crashes with   ```HTTP2::Error::StreamLimitExceeded``` error, looks like rpush is not setup http2 client correctly

**Comment by soulfly** [comment]
@Fivell are you able to provide more information about the crash?

**Comment by Fivell** [comment]
@soulfly , sorry,my bad,  rpush is not crashed it just error_description  from http2 stored to push_notification instance, however looks like should be fixed somehow still

**Comment by dorthwein** [comment]
Was JWT Support added to this?  Thanks

**Comment by vtm9** [comment]
Was JWT Support added to this? Thanks

**Pull request merged**